docs/dev/framework/csp.md

```diff
@@ -0,0 +1,229 @@
+```
+{{% /tab %}}
+{{% tab name="Twig" %}}
+```twig
+{% do csp_source('frame-src', 'https://www.youtube.com/embed/foobar123') %}
+```
+{{% /tab %}}
+{{% tab name="PHP Template" %}}
+```php
+<?php $this->addCspSource('frame-src', 'https://www.youtube.com/embed/foobar123') ?>
+```
+{{% /tab %}}
+{{< /tabs >}}
+
+{{% notice "note" %}}
+The method will automatically not add the source, if no source list had been set yet for the given directive (or its
+fallbacks). This behavior can be switched off via the `$autoIgnore` parameter.
+{{% /notice %}}
+
+
+### Retrieving a Nonce
+
+When employing a CSP directive that affects JavaScript or CSS (`default-src`, `script-src`, `style-src`) inline scripts
+will not be allowed anymore as it is recommended to _only_ integrate scripts via files. However, if your application
+still wants to use inline script or style elements it can still be allowed through the usage of [nonces][CSPNonce], 
+without having to resort to allowing `'unsafe-inline'` in your directives. Contao's default JavaScript templates for 
+instance still use inline JavaScript and thus will add a nonce when used. You can retrieve a nonce for `script-src` or 
+`style-src` like this:
+
+{{< tabs groupId="csp-methods" >}}
+{{% tab name="PHP" %}}
+```php
+$nonce = $cspHandler->getNonce('script-src');
+```
+{{% /tab %}}
+{{% tab name="Twig" %}}
+```twig
+<script{{ attrs().setIfExists('nonce', csp_nonce('script-src')) }}>
+```
+{{% /tab %}}
+{{% tab name="PHP Template" %}}
+```php
+<script<?= $this->attr()->setIfExists('nonce', $this->nonce('script-src')) ?>>
+```
+{{% /tab %}}
+{{< /tabs >}}
+
+{{% notice "note" %}}
+The method will automatically not return a nonce, if no source list had been set yet for the given directive (or its
+fallbacks).
+{{% /notice %}}
+
+This nonce can then be added as a `nonce="…"` attribute to your `<script>` or `<style>` tag respectively.
+
+
+### Adding Hashes
+
+As mentioned in the previous section, any directives that affect JavaScript or CSS will automatically disallow any
+inline scripts or inline styles. This also includes inline scripts via HTML attributes, e.g. `onclick="…"` or
+`style="…"`. It is highly recommended to move these styles and scripts into files instead. However, if for some reason
+this is not possible such inline scripts can still be allowed through [hashes][CSPHash]. A hash can be generated and
+added to a directive for a specific inline script or style like this:
+
+{{< tabs groupId="csp-methods" >}}
+{{% tab name="PHP" %}}
+```php
+$cspHandler->addHash('style-src', 'display:none');
+```
+{{% /tab %}}
+{{% tab name="Twig" %}}
+```twig
+{% do csp_hash('display:none') %}
+<div style="display:none">
+```
+{{% /tab %}}
+{{% tab name="PHP Template" %}}
+```php
+<div style="<?= $this->cspInlineStyle('display:none') ?>">
+```
+{{% /tab %}}
+{{< /tabs >}}
+
+{{% notice "note" %}}
+In order for inline styles to work in browsers that support CSP Level 3 you will also need to add `'unsafe-hashes'` to
```

Comment:
The hint that this is taken care of for you if you  use `csp_hash()` in Twig or `cspInlineStyle()` in PHP templates is missing. Maybe the hint should go to the tabs themselves? You only need to take care of it in the PHP controller for example.

Reply:
Updated in fb4d9fc5f2c5050b2413c62b00b3af4cc5eb08bb